ambient-mesh/migration/v1/gather-cluster-info.sh

```diff
@@ -0,0 +1,600 @@
+  fi
+
+  # Use jq to construct the metrics object and save to temporary file
+  jq -n \
+    --arg ns_name "$ns_name" \
+    --arg out_ns_name "$out_ns_name" \
+    --argjson pods "$pod_count" \
+    --argjson is_istio_injected "$is_istio_injected" \
+    --argjson reg_containers "$regular_containers" \
+    --argjson reg_cpu "$regular_cpu" \
+    --argjson reg_mem "$regular_mem" \
+    --argjson reg_cpu_actual "$regular_cpu_actual" \
+    --argjson reg_mem_actual "$regular_mem_actual" \
+    --argjson istio_containers "$istio_containers" \
+    --argjson istio_cpu "$istio_cpu" \
+    --argjson istio_mem "$istio_mem" \
+    --argjson istio_cpu_actual "$istio_cpu_actual" \
+    --argjson istio_mem_actual "$istio_mem_actual" \
+    "$jq_filter" > "$output_file"
+
+  if [ $? -ne 0 ]; then
+    echo "{\"status\": \"error\", \"message\": \"Failed to construct metrics JSON for namespace $ns_name\"}" > "$output_file"
+    return 1
+  fi
+}
+
+# Get count of namespaces
+TOTAL_NAMESPACES=$(kubectl --context="$CONTEXT" get ns -o json | jq -r '.items[] | .metadata.name' | wc -l)
+log_info "Found a total of $TOTAL_NAMESPACES namespace(s) to process in context $CONTEXT"
+
+out_ctx=$CONTEXT
+if [ "$OBFUSCATE_NAMES" = true ]; then
+  out_ctx=$(echo "$CONTEXT" | sha256sum | awk '{print $1}')
+fi
+
+# Set the cluster name in the JSON
+jq --arg name "$out_ctx" '.name = $name' cluster_info.json > tmp.json && mv tmp.json cluster_info.json
+
+# Check if metrics API is available for this cluster
+has_metrics=false
+if kubectl --context="$CONTEXT" top pod -A >/dev/null 2>&1; then
+  has_metrics=true
+  log_info "Metrics API available for cluster $CONTEXT"
+else
+  log_warn "Metrics API not available for cluster $CONTEXT"
+fi
+
+# Update the has_metrics field
+jq --argjson has_metrics "$has_metrics" '.has_metrics = $has_metrics' cluster_info.json > tmp.json && mv tmp.json cluster_info.json
+
+# get list of nodes
+nodes=$(kubectl --context="$CONTEXT" get nodes -o json | jq -r '.items[] | .metadata.name')
+if [ -z "$nodes" ]; then
+  log_warn "No nodes found for cluster $CONTEXT"
+else
+  # process each node
+  for node in $nodes; do
+    process_node_data "$node" "$CONTEXT" "$has_metrics"
+  done
+fi
+
+# Cache namespaces with istio injection
+namespaces=$(kubectl --context="$CONTEXT" get ns -o json | \
+  jq -r '.items[] | .metadata.name')
+
+if [ -z "$namespaces" ]; then
+  log_warn "No namespaces found in context $CONTEXT"
+else
+  # Reset namespace counter
+  CURRENT_NAMESPACE=0
+  TOTAL_NAMESPACES=$(echo "$namespaces" | wc -l)
+  update_progress
+
+  # Process each namespace in parallel with controlled concurrency
+  running=0
+  touch "$TEMP_DIR/warnings.log"
+
+  while IFS= read -r ns_name; do
+    # Wait if we've reached the maximum number of parallel processes
+    while [ $running -ge $MAX_PARALLEL ]; do
+      # Wait for a child process to finish
+      wait -n 2>/dev/null || true
+      # Count remaining background jobs
+      running=$(jobs -p | wc -l)
```

Comment:
I did a little testing, and it seems like this works correctly (on my mac). But I'm not totally confident that the behavior of `jobs` is consistent across environments, and I'm a little concerned that this could lead to deadlocks if the count is off for some reason.

Reply:
I've done more reading about this and I'm guessing it's probably fine?

Reply:
yeah, the safest bet was to move away from `jobs` and `wait -n` for a more manual batch process (manually update a counter and just using `wait`). makes it slower, but should hopefully avoid these issues